### PR TITLE
Feat/doublelink

### PR DESCRIPTION
**Pull request by zuoxiaodong0815**
Intra-line double link

### CODE REVIEW

File: libs/components/common/src/lib/text/plugins/DoubleLink.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    id: string;
+};
+
+export const DoubleLinkComponent = ({ attributes, children, element }: any) => {
+    const navigate = useNavigate();
+
+    const handleClickLinkText = useCallback(
+        (event: React.MouseEvent<HTMLAnchorElement>) => {
+            event.preventDefault();
+            event.stopPropagation();
```
Comment (lawvs) [inline libs/components/common/src/lib/text/plugins/DoubleLink.tsx:20]: Is this necessary?

File: libs/components/common/src/lib/text/plugins/DoubleLink.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+export type DoubleLinkElement = {
+    type: 'link';
+    workspaceId: string;
+    blockId: string;
+    children: Descendant[];
+    id: string;
+};
+
+export const DoubleLinkComponent = ({ attributes, children, element }: any) => {
```
Comment (lawvs) [inline libs/components/common/src/lib/text/plugins/DoubleLink.tsx:14]: Please declare the type explicitly

File: libs/components/common/src/lib/text/plugins/DoubleLink.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    return (
+        <span>
+            <PagesIcon style={{ verticalAlign: 'middle', height: '20px' }} />
+            <a
+                {...attributes}
+                style={{ cursor: 'pointer' }}
+                href={`/${element.workspaceId}/${element.blockId}`}
+            >
+                <span onClick={handleClickLinkText}>{children}</span>
```
Comment (lawvs) [inline libs/components/common/src/lib/text/plugins/DoubleLink.tsx:35]: It looks weird here

File: libs/components/common/src/lib/text/slate-utils.ts [hunk 4]
```diff
@@ -593,8 +597,13 @@ class SlateUtils {
 
         const fragmentChildren = firstFragment.children;
 
-        const textChildren: Text[] = [];
-        for (const child of fragmentChildren) {
+        const textChildren = [];
```
Comment (lawvs) [inline libs/components/common/src/lib/text/slate-utils.ts:600]: Please declare the type explicitly

File: libs/components/common/src/lib/text/slate-utils.ts [hunk 5]
```diff
@@ -593,8 +597,13 @@ class SlateUtils {
...
+        for (let i = 0; i < fragmentChildren.length; i++) {
+            const child = fragmentChildren[i];
+            if ('type' in child && child.type === 'link') {
+                i !== fragmentChildren.length - 1 && textChildren.push(child);
```
Comment (lawvs) [inline libs/components/common/src/lib/text/slate-utils.ts:604]: Please explain why the last element should be skipped

File: libs/components/editor-plugins/src/search/Search.tsx [hunk 6]
```diff
@@ -37,9 +36,7 @@ const query_blocks = (
     search: string,
     callback: (result: QueryResult) => void
 ) => {
-    (editor as BlockEditor)
-        .search(search)
-        .then(pages => callback(pages.filter(b => !!b.content)));
+    (editor as BlockEditor).search(search).then(pages => callback(pages));
```
Comment (lawvs) [inline libs/components/editor-plugins/src/search/Search.tsx:39]: why not return a promise directly?

File: libs/datasource/jwt/src/index.ts [hunk 7]
```diff
@@ -290,19 +289,22 @@ export class BlockClient<
...
+                this.search(part_of_title_or_content).flatMap(({ result }) =>
+                    result.map(async id => {
+                        const page = this._pageMapping.get(id as string);
+                        if (page) return page;
+                        const block = await this.get(id as BlockTypeKeys);
+                        return this.set_page(block);
+                    })
+                )
+            );
+            pages = [...new Set(promisedPages.filter((v): v is string => !!v))];
```
Comment (lawvs) [inline libs/datasource/jwt/src/index.ts:304]: This type guard is incorrect

File: libs/components/editor-plugins/src/menu/double-link-menu/DoubleLinkMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+                    editor.selection.currentSelectInfo.browserSelection
+                );
+                if (textSelection) {
+                    const { anchor } = textSelection;
+                    editor.blockHelper.setDoubleLinkSearchSlash(
+                        nextNodeId,
+                        anchor
+                    );
+                }
+            });
```
Comment (lawvs) [inline libs/components/editor-plugins/src/menu/double-link-menu/DoubleLinkMenu.tsx:171]: Please comment the purpose of the setTimeout

File: libs/components/editor-plugins/src/menu/double-link-menu/Container.tsx [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+            if (!types.includes(currentItem)) {
+                setNeedCheckIntoView(true);
+                if (types.length) {
+                    setCurrentItem(types[0]);
+                } else {
+                    setCurrentItem(undefined);
+                }
+            }
+        }
+    }, [types, currentItem]);
```
Comment (lawvs) [inline libs/components/editor-plugins/src/menu/double-link-menu/Container.tsx:75]: Please do not abuse useEffect

File: libs/components/editor-core/src/editor/block/async-block.ts [hunk 10]
```diff
@@ -495,4 +497,33 @@ export class AsyncBlock {
...
+                    const newItem = {
+                        ...item,
+                        children: children,
+                    };
+                    values.splice(i, 1, newItem);
+                }
+            }
+        }
+        return rawData;
+    }
```
Comment (lawvs) [inline libs/components/editor-core/src/editor/block/async-block.ts:528]:
```suggestion
    async updateDoubleLinkBlock(rawData: ReturnEditorBlock) {
      const values = rawData.properties?.text?.value || [];
      for (let i = 0; i < values.length; i++) {
        const item = values[i] as any;
        if (item.linkType !== "doubleLink") {
          continue;
        }
        const linkBlock = await this.services.load({
          workspace: item.workspaceId,
          id: item.blockId,
        });

        let children = linkBlock?.getProperties().text?.value || [];
        if (children.length === 1 && !children[0].text) {
          children = [{ text: "Untitled" }];
        }
        if (
          children.map((v) => v.text).join("") !==
          (item.children || []).map((v: any) => v.text).join("")
        ) {
          const newItem = {
            ...item,
            children,
          };
          values.splice(i, 1, newItem);
        }
      }
      return rawData;
    }
```

File: libs/components/editor-core/src/editor/block/async-block.ts [hunk 11]
```diff
@@ -154,13 +154,15 @@ export class AsyncBlock {
...
         this.raw_data = await this.filterPageInvalidChildren(this.raw_data);
+        this.raw_data = await this.updateDoubleLinkBlock(this.raw_data);
         const { workspace, id } = this.raw_data;
         this.unobserve = await this.services.observe(
             { workspace, id },
             async blockData => {
                 const oldData = this.raw_data;
                 this.raw_data = blockData;
                 this.raw_data = await this.filterPageInvalidChildren(blockData);
+                this.raw_data = await this.updateDoubleLinkBlock(this.raw_data);
```
Comment (lawvs) [inline libs/components/editor-core/src/editor/block/async-block.ts:165]: Is real sync necessary? I'm worried it will cost a lot of performances.

**Comment by lawvs** [comment]
I think the name `doubleLink` is not appropriate. Is there a better alternative? For example `internalLink` or `pageReference`?

**Review by darkskygit — APPROVED** [review]

**Pull request merged**